File: desktop/macos/Desktop/Sources/Bluetooth/Connections/PlaudDeviceConnection.swift
@preconcurrency import CoreBluetooth
import Foundation
import os.log

/// Device connection implementation for PLAUD NotePin devices
/// Ported from: omi/app/lib/services/devices/plaud_connection.dart
final class PlaudDeviceConnection: BaseDeviceConnection {

  // MARK: - Constants

  private enum Command {
    static let getBattery = 9
    static let startRecord = 20
    static let stopRecord = 23
    static let syncFileStart = 28
    static let stopSync = 30
  }

  // MARK: - Private Properties

  private let logger = Logger(subsystem: "me.omi.desktop", category: "PlaudDeviceConnection")

  private let commandOperations: DeviceOperationBroker<Int, Data>
  private let commandGate = UncorrelatedOperationGate<Int>()
  private let commandQueue = DeviceCommandQueue()
  private var notificationSubscription: Task<Void, Never>?
  private var sessionId: Int?
  private var recordingSetupAttempted = false
  private lazy var audioStreamController = DeviceAudioStreamController(
    start: { [weak self] in
      guard let self else { throw DeviceTransportError.disposed }
      try await self.setupRecordingSession()
    },
    stop: { [weak self] in
      guard let self else { return }
      try await self.stopRecordingSession()
    }
  )

  // MARK: - Initialization

  override init(
    device: BtDevice,
    transport: DeviceTransport,
    operationClock: any DeviceOperationClock = ContinuousDeviceOperationClock()
  ) {
    self.commandOperations = DeviceOperationBroker(clock: operationClock)
    super.init(
      device: device,
      transport: transport,
      operationClock: operationClock
    )
  }

  // MARK: - Connection

  override func prepareDeviceAfterConnect() async throws {
    try await operationClock.sleep(for: .seconds(2))

    startNotificationListener()
  }

  override func teardownDevice() async {
    await audioStreamController.finish()
    await commandQueue.close()

    notificationSubscription?.cancel()
    notificationSubscription = nil
    await commandOperations.cancelAll(reason: .disconnected)
    commandGate.reset()
    sessionId = nil
  }

  // MARK: - Notification Handling

  private func startNotificationListener() {
    let stream = transport.getCharacteristicStream(
      serviceUUID: DeviceUUIDs.PLAUD.service,
      characteristicUUID: DeviceUUIDs.PLAUD.notifyCharacteristic
    )

    notificationSubscription = Task {
      do {
        for try await data in stream {
          handleNotification(Array(data))
        }
      } catch {
        logger.debug("Notification stream ended: \(error.localizedDescription)")
      }
    }
  }

  /// Parses one physical notification. Kept internal so adapter-correlation
  /// behavior can be verified without a real CoreBluetooth peripheral.
  func handleNotification(_ data: [UInt8]) {
    guard !data.isEmpty else { return }

    if data[0] == 2 {
      // Audio data packet
      if let chunk = parseAudioChunk(Array(data.dropFirst())) {
        audioStreamController.yield(Data(chunk))
      }
    } else if data.count >= 3 {
      // Command response
      let cmdId = Int(data[1]) | (Int(data[2]) << 8)
      let payload = data.count > 3 ? Array(data.dropFirst(3)) : []

      completeCommand(commandID: cmdId, payload: payload)
    }
  }

  private func completeCommand(commandID: Int, payload: [UInt8]) {
    guard let handle = commandGate.takeHandleForCallback(key: commandID) else { return }
    Task {
      await commandOperations.succeed(
        handle: handle,
        value: Data(payload)
      )
    }
  }

  private func parseAudioChunk(_ payload: [UInt8]) -> [UInt8]? {
    guard payload.count >= 9 else { return nil }

    let position = toInt32(Array(payload[4..<8]))
    if position == 0xFFFF_FFFF { return nil }  // End marker

    let length = Int(payload[8])
    guard payload.count >= 9 + length else { return nil }

    return Array(payload[9..<(9 + length)])
  }

  // MARK: - Command Sending

  func sendCommand(cmdId: Int, payload: [UInt8]) async throws -> [UInt8]? {
    try await commandQueue.run { [weak self] in
      guard let self else { throw DeviceTransportError.disposed }
      return try await self.sendCommandSerially(cmdId: cmdId, payload: payload)
    }
  }

  private func sendCommandSerially(cmdId: Int, payload: [UInt8]) async throws -> [UInt8]? {
    guard commandGate.canStart(cmdId) else {
      throw DeviceConnectionError.operationFailed(
        "A previous command response is ambiguous; reconnect the device"
      )
    }
    let command: [UInt8] = [1, UInt8(cmdId & 0xFF), UInt8((cmdId >> 8) & 0xFF)] + payload

    do {
      let response = try await commandOperations.perform(
        key: cmdId,
        timeout: .seconds(10),
        onTerminal: { [weak self] handle, termination in
          self?.commandGate.terminal(handle: handle, termination: termination)
        }
      ) { [weak self] handle in
        guard let self else { throw DeviceTransportError.disposed }
        guard self.commandGate.register(handle) else {
          throw DeviceConnectionError.operationFailed(
            "Command callback identity is no longer trustworthy"
          )
        }
        try await self.transport.writeCharacteristic(
          data: Data(command),
          serviceUUID: DeviceUUIDs.PLAUD.service,
          characteristicUUID: DeviceUUIDs.PLAUD.writeCharacteristic,
          withResponse: true
        )
      }
      return Array(response)
    } catch DeviceOperationBrokerError.timedOut {
      return nil
    }
  }

  // MARK: - Recording Control

  private func startRecord() async throws -> (sessionId: Int, startTime: Int)? {
    let payload = toBytes32(1) + toBytes32(0) + toBytes32(0)
    guard let response = try await sendCommand(cmdId: Command.startRecord, payload: payload),
      response.count >= 10
    else {
      return nil
    }

    return (
      sessionId: toInt32(Array(response[0..<4])),
      startTime: toInt32(Array(response[4..<8]))
    )
  }

  private func stopRecord(sessionId: Int) async throws {
    let payload = toBytes32(sessionId) + toBytes32(0)
    guard try await sendCommand(cmdId: Command.stopRecord, payload: payload) != nil else {
      throw DeviceConnectionError.operationFailed(
        "PLAUD did not acknowledge STOP_RECORD"
      )
    }
  }

  private func startSync(sessionId: Int, start: Int) async throws -> Bool {
    let payload = toBytes64(sessionId) + toBytes64(start) + toBytes64(0x7FFF_FFFF)
    let response = try await sendCommand(cmdId: Command.syncFileStart, payload: payload)
    return response != nil
  }

  private func stopSync() async throws {
    let command: [UInt8] = [1, UInt8(Command.stopSync & 0xFF), UInt8((Command.stopSync >> 8) & 0xFF), 1]
    try await commandQueue.run { [weak self] in
      guard let self else { throw DeviceTransportError.disposed }
      try await self.transport.writeCharacteristic(
        data: Data(command),
        serviceUUID: DeviceUUIDs.PLAUD.service,
        characteristicUUID: DeviceUUIDs.PLAUD.writeCharacteristic,
        withResponse: true
      )
    }
  }

  private func setupRecordingSession() async throws {
    recordingSetupAttempted = true
    let maxRetries = 3
    var lastError: Error?

    for attempt in 0..<maxRetries {
      try Task.checkCancellation()
      do {
        if attempt > 0 {
          logger.debug("Retry attempt \(attempt)/\(maxRetries)")
          try await operationClock.sleep(for: .seconds(attempt))
        }

        try await stopRecord(sessionId: 0)
        try await operationClock.sleep(for: .milliseconds(500))

        guard let recordInfo = try await startRecord() else { continue }

        sessionId = recordInfo.sessionId

        try await operationClock.sleep(for: .seconds(1))

        if try await startSync(sessionId: recordInfo.sessionId, start: recordInfo.startTime) {
          logger.debug("Recording session setup successful")
          return
        }
      } catch {
        guard !Task.isCancelled else { throw CancellationError() }
        lastError = error
        logger.debug("Setup error (attempt \(attempt + 1)): \(error.localizedDescription)")
      }
    }

    throw lastError
      ?? DeviceConnectionError.operationFailed(
        "PLAUD recording session setup failed"
      )
  }

  private func stopRecordingSession() async throws {
    guard recordingSetupAttempted else { return }
    recordingSetupAttempted = false
    let sessionToStop = sessionId ?? 0
    var failures: [String] = []
    do {
      try await stopSync()
    } catch {
      failures.append(error.localizedDescription)
    }
    do {
      try await stopRecord(sessionId: sessionToStop)
    } catch {
      failures.append(error.localizedDescription)
    }
    sessionId = nil

    guard failures.isEmpty else {
      let message = failures.joined(separator: "; ")
      logger.debug("Error stopping recording session: \(message)")
      await transport.disconnect()
      throw DeviceConnectionError.operationFailed(message)
    }
  }

  // MARK: - Battery

  override func getBatteryLevel() async -> Int {
    guard await isConnected() else { return -1 }

    do {
      guard let response = try await sendCommand(cmdId: Command.getBattery, payload: []),
        response.count >= 2
      else {
        return -1
      }

      // Response format: [is_charging, battery_level]
      let batteryLevel = Int(response[1])
      let isCharging = response[0] != 0
      logger.debug("Battery: \(batteryLevel)% \(isCharging ? "(Charging)" : "")")
      return batteryLevel
    } catch {
      logger.debug("Error retrieving battery level: \(error.localizedDescription)")
      return -1
    }
  }

  /// Get battery state including charging status
  func getBatteryState() async -> (level: Int, isCharging: Bool)? {
    guard await isConnected() else { return nil }

    do {
      guard let response = try await sendCommand(cmdId: Command.getBattery, payload: []),
        response.count >= 2
      else {
        return nil
      }

      return (level: Int(response[1]), isCharging: response[0] != 0)
    } catch {
      logger.debug("Error getting battery state: \(error.localizedDescription)")
      return nil
    }
  }

  override func getBatteryLevelStream() -> AsyncThrowingStream<Int, Error> {
    // PLAUD uses command-based battery retrieval, poll every 60 seconds
    return AsyncThrowingStream { [weak self] continuation in
      let pollingTask = Task { [weak self] in
        guard let self = self else {
          continuation.finish()
          return
        }

        // Get initial battery level
        let initialLevel = await self.getBatteryLevel()
        if initialLevel >= 0 {
          continuation.yield(initialLevel)
        }

        // Poll every 60 seconds
        var lastLevel = initialLevel
        while await self.isConnected() {
          do {
            try await self.operationClock.sleep(for: .seconds(60))
          } catch {
            break
          }
          guard !Task.isCancelled else { break }

          let level = await self.getBatteryLevel()
          if level >= 0 && level != lastLevel {
            lastLevel = level
            continuation.yield(level)
          }
        }

        continuation.finish()
      }
      continuation.onTermination = { @Sendable _ in
        pollingTask.cancel()
      }
    }
  }

  // MARK: - Audio

  override func getAudioCodec() async -> BleAudioCodec {
    .opusFS320
  }

  override func getAudioStream() -> AsyncThrowingStream<Data, Error> {
    let rawStream = audioStreamController.makeStream()
    return AsyncThrowingStream { continuation in
      let chunkingTask = Task { @MainActor in
        var buffer = [UInt8]()
        do {
          for try await data in rawStream {
            buffer.append(contentsOf: data)
            while buffer.count >= 80 {
              continuation.yield(Data(buffer.prefix(80)))
              buffer.removeFirst(80)
            }
          }
          if !buffer.isEmpty {
            continuation.yield(Data(buffer))
          }
          continuation.finish()
        } catch {
          continuation.finish(throwing: error)
        }
      }
      continuation.onTermination = { @Sendable _ in
        chunkingTask.cancel()
      }
    }
  }

  // MARK: - Unsupported Features

  override func getButtonState() async -> [UInt8] { [] }
  override func getButtonStream() -> AsyncThrowingStream<[UInt8], Error> {
    AsyncThrowingStream { $0.finish() }
  }

  override func hasPhotoStreaming() async -> Bool { false }
  override func getAccelerometerStream() -> AsyncThrowingStream<AccelerometerData, Error> {
    AsyncThrowingStream { $0.finish() }
  }

  override func getFeatures() async -> OmiFeatures { [] }

  // MARK: - Byte Conversion Helpers

  private func toBytes32(_ value: Int) -> [UInt8] {
    [
      UInt8(value & 0xFF),
      UInt8((value >> 8) & 0xFF),
      UInt8((value >> 16) & 0xFF),
      UInt8((value >> 24) & 0xFF),
    ]
  }

  private func toBytes64(_ value: Int) -> [UInt8] {
    [
      UInt8(value & 0xFF),
      UInt8((value >> 8) & 0xFF),
      UInt8((value >> 16) & 0xFF),
      UInt8((value >> 24) & 0xFF),
      UInt8((value >> 32) & 0xFF),
      UInt8((value >> 40) & 0xFF),
      UInt8((value >> 48) & 0xFF),
      UInt8((value >> 56) & 0xFF),
    ]
  }

  private func toInt32(_ bytes: [UInt8]) -> Int {
    guard bytes.count >= 4 else { return 0 }
    return Int(bytes[0]) | (Int(bytes[1]) << 8) | (Int(bytes[2]) << 16) | (Int(bytes[3]) << 24)
  }
}

File: desktop/windows/src/renderer/src/device/connections/plaudDeviceConnection.ts
/**
 * PLAUD NotePin connection — Windows port of macOS
 * Connections/PlaudDeviceConnection.swift. One notify characteristic carries
 * both command responses and audio, demuxed by the leading type byte; the
 * recording session is started with retries on the first audio subscriber
 * and stopped on the last.
 */

import { PLAUD_UUIDS } from '../protocol/uuids'
import type { BleAudioCodec } from '../protocol/deviceTypes'
import {
  DeviceOperationBroker,
  DeviceOperationBrokerError,
  UncorrelatedOperationGate,
  type DeviceOperationClock
} from '../session/deviceOperationBroker'
import { DeviceCommandQueue } from '../session/deviceCommandQueue'
import { DeviceAudioStreamController } from '../session/deviceAudioStreamController'
import {
  BaseDeviceConnection,
  DeviceConnectionError,
  finishedStream,
  type AccelerometerData,
  type StreamSubscriber,
  type StreamSubscription
} from './deviceConnection'
import { FixedSizeRechunker } from './fixedSizeRechunker'
import type { DeviceTransport } from '../transport/deviceTransport'
import type { BtDevice } from '../protocol/btDevice'

const CMD_GET_BATTERY = 9
const CMD_START_RECORD = 20
const CMD_STOP_RECORD = 23
const CMD_SYNC_FILE_START = 28
const CMD_STOP_SYNC = 30

const PACKET_TYPE_COMMAND = 0x01
const PACKET_TYPE_AUDIO = 0x02
const COMMAND_TIMEOUT_MS = 10_000
const POST_CONNECT_SETTLE_MS = 2_000
const OUTPUT_CHUNK_SIZE = 80
const END_OF_STREAM_POSITION = 0xffff_ffff
const SYNC_END_SENTINEL = 0x7fff_ffff
const BATTERY_POLL_INTERVAL_MS = 60_000
const SETUP_MAX_RETRIES = 3
const SETUP_STOP_TO_START_GAP_MS = 500
const SETUP_START_TO_SYNC_GAP_MS = 1_000

const commandKey = (cmdId: number): string => `plaud-cmd:${cmdId}`

const toBytes32 = (value: number): Uint8Array =>
  Uint8Array.from([
    value & 0xff,
    (value >>> 8) & 0xff,
    (value >>> 16) & 0xff,
    (value >>> 24) & 0xff
  ])

const toBytes64 = (value: number): Uint8Array => {
  const big = BigInt(value)
  const out = new Uint8Array(8)
  for (let i = 0; i < 8; i += 1) {
    out[i] = Number((big >> BigInt(8 * i)) & 0xffn)
  }
  return out
}

const toInt32LE = (data: Uint8Array, offset: number): number => {
  const value =
    (data[offset] |
      (data[offset + 1] << 8) |
      (data[offset + 2] << 16) |
      (data[offset + 3] << 24)) >>>
    0
  return value > 0x7fff_ffff ? value - 0x1_0000_0000 : value
}

export interface PlaudBatteryState {
  level: number
  isCharging: boolean
}

export class PlaudDeviceConnection extends BaseDeviceConnection {
  private readonly commandQueue = new DeviceCommandQueue()
  private readonly commandBroker: DeviceOperationBroker
  private readonly commandGate = new UncorrelatedOperationGate()
  private readonly audioController: DeviceAudioStreamController
  private sessionId: number | null = null
  private setupAttempted = false

  constructor(args: {
    device: BtDevice
    transport: DeviceTransport
    clock?: DeviceOperationClock
  }) {
    super(args)
    this.commandBroker = new DeviceOperationBroker(this.clock)
    this.audioController = new DeviceAudioStreamController({
      start: () => this.setupRecordingSession(),
      stop: () => this.stopRecordingSession()
    })
  }

  protected override async prepareDeviceAfterConnect(): Promise<void> {
    await this.settle(POST_CONNECT_SETTLE_MS)
    this.ensureLifecycleIsActive()
    this.transport.subscribeCharacteristic(PLAUD_UUIDS.service, PLAUD_UUIDS.notifyCharacteristic, {
      onData: (data) => this.handleNotification(data),
      onFinish: (error) => {
        void this.audioController.finish(error)
      }
    })
  }

  protected override async teardownDevice(): Promise<void> {
    await this.audioController.finish(null)
    this.commandBroker.cancelAll('disconnected')
    this.commandGate.reset()
    await this.commandQueue.close()
  }

  // --- command channel -------------------------------------------------------

  /** Command frame: leading type byte 1, LE 16-bit id, payload. Timeout
   *  resolves null; a poisoned command identity throws. */
  private sendCommand(cmdId: number, payload: Uint8Array): Promise<Uint8Array | null> {
    return this.commandQueue.run(async () => {
      const key = commandKey(cmdId)
      const handle = this.commandGate.register(key)
      if (handle === null) {
        throw DeviceConnectionError.operationFailed(
          'A previous command response is ambiguous; reconnect the device'
        )
      }
      const frame = new Uint8Array(3 + payload.length)
      frame[0] = PACKET_TYPE_COMMAND
      frame[1] = cmdId & 0xff
      frame[2] = (cmdId >> 8) & 0xff
      frame.set(payload, 3)
      try {
        return await this.commandBroker.perform<Uint8Array>({
          key,
          timeoutMs: COMMAND_TIMEOUT_MS,
          onTerminal: (termination) => this.commandGate.terminal(handle, termination),
          start: () =>
            this.transport.writeCharacteristic({
              serviceUuid: PLAUD_UUIDS.service,
              characteristicUuid: PLAUD_UUIDS.writeCharacteristic,
              data: frame,
              withResponse: true
            })
        })
      } catch (error) {
        if (error instanceof DeviceOperationBrokerError && error.kind === 'timedOut') {
          return null
        }
        throw this.mapTransportError(error)
      }
    })
  }

  private handleNotification(data: Uint8Array): void {
    if (data.length === 0) return
    if (data[0] === PACKET_TYPE_AUDIO) {
      this.parseAudioChunk(data.subarray(1))
      return
    }
    if (data.length >= 3) {
      const cmdId = data[1] | (data[2] << 8)
      const key = commandKey(cmdId)
      if (this.commandGate.takeHandleForCallback(key) === null) return
      this.commandBroker.succeed(key, data.slice(3))
    }
  }

  private parseAudioChunk(payload: Uint8Array): void {
    // [0..3] session id (unused), [4..7] LE position, [8] length, [9..] Opus.
    if (payload.length < 9) return
    const position =
      (payload[4] | (payload[5] << 8) | (payload[6] << 16) | (payload[7] << 24)) >>> 0
    if (position === END_OF_STREAM_POSITION) return
    const length = payload[8]
    if (payload.length < 9 + length) return
    this.audioController.yieldFrame(payload.slice(9, 9 + length))
  }

  // --- recording session -----------------------------------------------------

  private async startRecord(): Promise<{ sessionId: number; startTime: number } | null> {
    const payload = new Uint8Array(12)
    payload.set(toBytes32(1), 0)
    const response = await this.sendCommand(CMD_START_RECORD, payload)
    if (response === null || response.length < 10) return null
    return { sessionId: toInt32LE(response, 0), startTime: toInt32LE(response, 4) }
  }

  private async stopRecord(sessionId: number): Promise<void> {
    const payload = new Uint8Array(8)
    payload.set(toBytes32(sessionId), 0)
    const response = await this.sendCommand(CMD_STOP_RECORD, payload)
    if (response === null) {
      throw DeviceConnectionError.operationFailed('PLAUD did not acknowledge STOP_RECORD')
    }
  }

  private async startSync(sessionId: number, start: number): Promise<boolean> {
    const payload = new Uint8Array(24)
    payload.set(toBytes64(sessionId), 0)
    payload.set(toBytes64(start), 8)
    payload.set(toBytes64(SYNC_END_SENTINEL), 16)
    const response = await this.sendCommand(CMD_SYNC_FILE_START, payload)
    return response !== null
  }

  private async stopSync(): Promise<void> {
    // Raw frame with no response correlation, still serialized on the queue.
    const frame = Uint8Array.from([
      PACKET_TYPE_COMMAND,
      CMD_STOP_SYNC & 0xff,
      (CMD_STOP_SYNC >> 8) & 0xff,
      0x01
    ])
    await this.commandQueue.run(() =>
      this.transport.writeCharacteristic({
        serviceUuid: PLAUD_UUIDS.service,
        characteristicUuid: PLAUD_UUIDS.writeCharacteristic,
        data: frame,
        withResponse: true
      })
    )
  }

  private async setupRecordingSession(): Promise<void> {
    this.setupAttempted = true
    let lastError: Error | null = null
    for (let attempt = 0; attempt < SETUP_MAX_RETRIES; attempt += 1) {
      if (attempt > 0) {
        await this.settle(attempt * 1_000)
      }
      try {
        await this.stopRecord(0)
        await this.settle(SETUP_STOP_TO_START_GAP_MS)
        const record = await this.startRecord()
        if (record === null) {
          lastError = DeviceConnectionError.operationFailed(
            'PLAUD did not acknowledge START_RECORD'
          )
          continue
        }
        this.sessionId = record.sessionId
        await this.settle(SETUP_START_TO_SYNC_GAP_MS)
        if (!(await this.startSync(record.sessionId, record.startTime))) {
          lastError = DeviceConnectionError.operationFailed(
            'PLAUD did not acknowledge SYNC_FILE_START'
          )
          continue
        }
        return
      } catch (error) {
        lastError = error instanceof Error ? error : new Error(String(error))
      }
    }
    throw lastError ?? DeviceConnectionError.operationFailed('PLAUD recording session setup failed')
  }

  private async stopRecordingSession(): Promise<void> {
    if (!this.setupAttempted) return
    const failures: string[] = []
    try {
      await this.stopSync()
    } catch (error) {
      failures.push(error instanceof Error ? error.message : String(error))
    }
    try {
      await this.stopRecord(this.sessionId ?? 0)
    } catch (error) {
      failures.push(error instanceof Error ? error.message : String(error))
    }
    if (failures.length > 0) {
      await this.transport.disconnect()
      throw DeviceConnectionError.operationFailed(failures.join('; '))
    }
  }

  // --- audio -----------------------------------------------------------------

  override async getAudioCodec(): Promise<BleAudioCodec> {
    return 'opusFS320'
  }

  override getAudioStream(subscriber: StreamSubscriber<Uint8Array>): StreamSubscription {
    const rechunker = new FixedSizeRechunker(OUTPUT_CHUNK_SIZE)
    const subscription = this.audioController.subscribe({
      onFrame: (frame) => {
        for (const chunk of rechunker.push(frame)) {
          subscriber.onValue(chunk)
        }
      },
      onFinish: (error) => {
        const tail = rechunker.flush()
        if (tail !== null) subscriber.onValue(tail)
        subscriber.onFinish(error)
      }
    })
    return { cancel: () => subscription.cancel() }
  }

  // --- battery (byte order reversed vs Bee) ----------------------------------

  async getBatteryState(): Promise<PlaudBatteryState | null> {
    try {
      const response = await this.sendCommand(CMD_GET_BATTERY, new Uint8Array(0))
      if (response === null || response.length < 2) return null
      return { level: response[1], isCharging: response[0] !== 0 }
    } catch {
      return null
    }
  }

  override async getBatteryLevel(): Promise<number> {
    const state = await this.getBatteryState()
    return state?.level ?? -1
  }

  override getBatteryLevelStream(subscriber: StreamSubscriber<number>): StreamSubscription {
    const abort = new AbortController()
    this.sessionAbort.signal.addEventListener('abort', () => abort.abort(), { once: true })
    void (async () => {
      let lastLevel: number | null = null
      for (;;) {
        // A poisoned command id can never produce another response, so polling
        // it forever would just report a stale level indefinitely.
        if (this.commandGate.isPoisoned(commandKey(CMD_GET_BATTERY))) {
          subscriber.onFinish(
            DeviceConnectionError.operationFailed(
              'PLAUD battery reporting is ambiguous; reconnect the device'
            )
          )
          return
        }
        const state = await this.getBatteryState()
        if (abort.signal.aborted) return
        if (state !== null && state.level !== lastLevel) {
          lastLevel = state.level
          subscriber.onValue(state.level)
        }
        const outcome = await this.clock.sleep(BATTERY_POLL_INTERVAL_MS, abort.signal)
        if (outcome !== 'elapsed') return
      }
    })()
    return { cancel: () => abort.abort() }
  }

  // --- unsupported surfaces --------------------------------------------------

  override async getButtonState(): Promise<Uint8Array> {
    return new Uint8Array(0)
  }

  override getButtonStream(subscriber: StreamSubscriber<Uint8Array>): StreamSubscription {
    return finishedStream(subscriber)
  }

  override getAccelerometerStream(
    subscriber: StreamSubscriber<AccelerometerData>
  ): StreamSubscription {
    return finishedStream(subscriber)
  }

  override async getFeatures(): Promise<[]> {
    return []
  }
}
